File: problem_3.cpp
/*
 *                  Spiral Matrix
 *
 * Leetcode : https://leetcode.com/problems/spiral-matrix/description/
 * Time complexity : O(m * n)
 * Space complexity : O(m * n)
 * Approach :
 *      Using recursion
 */

#include<iostream>
#include<vector>

using namespace std;



vector<int> spiral_matrix(vector<vector<int> >& matrix) {

}

int main() {
    vector<vector<int> > matrix;
    // Creating the first row [1, 2, 3]
    vector<int> row1;
    row1.push_back(1);
    row1.push_back(2);
    row1.push_back(3);
    matrix.push_back(row1);

    // Creating the second row [4, 5, 6]
    vector<int> row2;
    row2.push_back(4);
    row2.push_back(5);
    row2.push_back(6);
    matrix.push_back(row2);

    // Creating the third row [7, 8, 9]
    vector<int> row3;
    row3.push_back(7);
    row3.push_back(8);
    row3.push_back(9);
    matrix.push_back(row3);

    int m = matrix.size();
    int n = matrix[0].size();
    vector<int> result(m*n);

    result = spiral_matrix(matrix);

    for(int i = 0; i < result.size(); i++) {
        cout << result[i] << " ";
    }

    return 0;
}

File: problem3.java
// Time Complexity :O(m*n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :YES
// Any problem you faced while coding this :NO


// Your code here along with comments explaining your approach
//accessing every element by handling i,j by motions in every direction and changing the value
// of accessed element to 200 to avoid repeatation ==> direction logic in code.


class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int m = matrix.length;
        int n = matrix[0].length;
        List<Integer> result = new ArrayList<>();
        int i = 0;
        int j = 0;
        int dir = 0;
        for(int index=0; index<m*n; index++){
            result.add(matrix[i][j]);
            matrix[i][j] = 200;
            //right direction dir-0
            if(dir==0){
                if(j==n-1 || matrix[i][j+1] == 200){
                    // if at first row rightmost or left to element that was accessed while going down
                    dir = 3;
                    i++;
                }else{
                    j++;
                }
            }else if(dir==1){
                //left
                if(j==0 || matrix[i][j-1]==200){
                     // if at last row leftmost or right to element that was accessed while going up
                    dir = 2;
                    i--;
                }
                else{
                    j--;
                }
            }else if(dir==2){
                //up
                if(i==0||matrix[i-1][j]==200){
                     // if at first row topmost or below of element that was accessed while going right
                    dir = 0;
                    j++;
                }else{
                    i--;
                }
            }else{
                //down
                if(i==m-1||matrix[i+1][j]==200){
                     // if at last row downmost or above the element that was accessed while going left
                    dir = 1;
                    j--;
                }else{
                    i++;
                }
            }
        }
        return result;
    }
}